src/DessserQCheck.ml

```diff
@@ -173,61 +173,45 @@ let shrink_mac_type mt =
   loop to_simplest
 
 let rec shrink_value_type =
+  let open Iter in
   let vt_of_mn mn = mn.T.vtyp in
-  let shrink_fields mns make_typ f =
-    Array.iter (fun (_, mn) -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-    let shrink_mns =
-      let shrink (fn, mn) =
-        Iter.map (fun mn -> fn, mn) (shrink_maybe_nullable mn) in
-      Shrink.filter (fun mns -> Array.length mns > 1)
-        (Shrink.array ~shrink) mns |>
-      Iter.map make_typ in
-    shrink_mns f in
+  let shrink_fields f =
+    (Shrink.pair Shrink.nil shrink_maybe_nullable) f in
   function
-  | T.Unknown | T.Ext _ ->
-      Iter.empty
-  | T.Base Unit ->
-      Iter.empty
-  | T.Base mt ->
-      (fun f ->
-        shrink_mac_type mt (fun mt -> f (T.Base mt)))
-  | T.Usr _ ->
-      Iter.empty
+  | T.Unknown | T.Ext _ -> empty
+  | T.Base Unit -> empty
+  | T.Base mt -> shrink_mac_type mt >|= T.base
+  | T.Usr _ -> empty
   | T.Vec (dim, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f mn.vtyp ;
-          f (T.Vec (dim, mn))))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.vec dim) <+> (smn >|= vt_of_mn)
   | T.Lst mn ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Lst mn) ;
-          f mn.vtyp))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.lst) <+> (smn >|= vt_of_mn)
   | T.Set (st, mn) ->
-      (fun f ->
-        shrink_maybe_nullable mn (fun mn ->
-          f (T.Set (st, mn)) ;
-          f mn.vtyp))
+      let smn = shrink_maybe_nullable mn in
+      (smn >|= T.set st) <+> (smn >|= vt_of_mn)
   | T.Tup mns ->
-      (fun f ->
-        Array.iter (fun mn -> shrink_maybe_nullable mn (f % vt_of_mn)) mns ;
-        let shrink_mns =
-          Shrink.filter (fun mns -> Array.length mns > 1)
-            (Shrink.array ~shrink:shrink_maybe_nullable) mns |>
-          Iter.map (fun mns -> T.Tup mns) in
-        shrink_mns f)
+      let smns = Shrink.array ~shrink:shrink_maybe_nullable mns in
+      filter (fun mn -> Array.length mn > 1) smns >|= T.tup
+       <+>
+      (of_array mns >|= vt_of_mn)
```

Comment:
Why not drink the kool-aid down to the very last drop and also use the pipe operator here?
Also, let's keep `mns` for the array of maybe nullables rather than [mn].
Also, same note about order of iteration.

Reply:
Ok